ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DownloadHeadersStep.java

```diff
@@ -85,23 +95,52 @@ private CompletableFuture<List<BlockHeader>> downloadHeaders(final SyncTargetRan
               protocolSchedule,
               protocolContext,
               ethContext,
+              peerTaskExecutor,
+              synchronizerConfiguration,
               range.getEnd(),
               range.getSegmentLengthExclusive(),
               validationPolicy,
               metricsSystem)
           .run();
     } else {
       LOG.debug("Downloading headers starting from {}", range.getStart().getNumber());
-      return GetHeadersFromPeerByHashTask.startingAtHash(
-              protocolSchedule,
-              ethContext,
-              range.getStart().getHash(),
-              range.getStart().getNumber(),
-              headerRequestSize,
-              metricsSystem)
-          .assignPeer(range.getSyncTarget())
-          .run()
-          .thenApply(PeerTaskResult::getResult);
+      if (synchronizerConfiguration.isPeerTaskSystemEnabled()) {
+        return ethContext
+            .getScheduler()
+            .scheduleServiceTask(
+                () -> {
+                  LOG.info("Getting headers using peer task system");
+                  GetHeadersFromPeerTask task =
+                      new GetHeadersFromPeerTask(
+                          range.getStart().getHash(),
+                          range.getStart().getNumber(),
+                          headerRequestSize,
+                          0,
+                          GetHeadersFromPeerTask.Direction.FORWARD,
+                          protocolSchedule);
+                  PeerTaskExecutorResult<List<BlockHeader>> taskResult =
+                      peerTaskExecutor.executeAgainstPeer(task, range.getSyncTarget());
+                  if (taskResult.responseCode() != PeerTaskExecutorResponseCode.SUCCESS
+                      || taskResult.result().isEmpty()) {
+                    LOG.info("Unable to download headers for range {}", range);
+                    return CompletableFuture.failedFuture(
+                        new RuntimeException("Unable to download headers for range " + range));
+                  }
+                  LOG.info("Successfully downloaded headers");
+                  return CompletableFuture.completedFuture(taskResult.result().get());
```

Comment:
It is a common use case, but I don't like the idea of making PeerTaskExecutor return CompletableFutures when it's not doing any multi-threading. Avoiding simplification also allows us to include context exception messages which otherwise wouldn't be possible.